test/compound-rate-keeper/CompoundRateKeeperV2.test.js

```diff
@@ -0,0 +1,234 @@
+      await setNextBlockTime(10);
+      await crk.setAnnualPercent(decimal(1.1).toString());
+
+      assert.equal(fromDecimal((await crk.getPotentialCompoundRate(10 + 31536000)).toString()), "1.1");
+
+      await setNextBlockTime(10 + 31536000);
+      await crk.setAnnualPercent(decimal(1.2).toString());
+
+      assert.equal(fromDecimal((await crk.getPotentialCompoundRate(10 + 3 * 31536000)).toString()), "1.584");
+
+      await setNextBlockTime(10 + 3 * 31536000);
+      await crk.setAnnualPercent(decimal(1.5).toString());
+
+      assert.equal(fromDecimal((await crk.getPotentialCompoundRate(10 + 6 * 31536000)).toString()), "5.346");
+
+      await setNextBlockTime(10 + 6 * 31536000);
+      await crk.setAnnualPercent(decimal(1.05).toString());
+
+      assert.equal(fromDecimal((await crk.getPotentialCompoundRate(10 + 7 * 31536000)).toString()), "5.6133");
+    });
+
+    it("check max timestamp for 10%", async function () {
+      await checkByParams(1.1);
+    });
+
+    it("check max timestamp for 25%", async function () {
+      await checkByParams(1.25);
+    });
+
+    it("check max timestamp for 50%", async function () {
+      await checkByParams(1.5);
+    });
+
+    it("check max timestamp for 100%", async function () {
+      await checkByParams(2);
+    });
+
+    it("check max timestamp for 200%", async function () {
+      await checkByParams(3);
+    });
+
+    it("check max timestamp for 500%", async function () {
+      await checkByParams(6);
+    });
+
+    it("check max timestamp for 1000%", async function () {
+      await checkByParams(11);
+    });
+  });
+
+  async function checkByParams(annualPercent) {
+    await setNextBlockTime(10);
+    await crk.setAnnualPercent(decimal(annualPercent).toString());
+
+    let calcRate;
+    let expectRate;
+    for (let y = 3; y <= 3600; y += 3) {
+      try {
+        calcRate = (await crk.getPotentialCompoundRate(10 + y * 2628000)).toString();
+
+        const capitalizationPeriodsNum = Math.floor(y / 12);
+        const capitalizationRate = toBN(annualPercent).pow(capitalizationPeriodsNum);
+        const leftRate = toBN(y % 12)
+          .multipliedBy(toBN(annualPercent).minus(1))
+          .dividedBy(12)
+          .plus(1);
+        expectRate = toBN(decimal(capitalizationRate.multipliedBy(leftRate))).toFixed(0);
+
+        if (toBN(expectRate).isGreaterThan(maxRate)) {
+          assert.equal(calcRate, maxRate);
+        } else {
+          assert.closeTo(toBN(calcRate).minus(expectRate).toNumber(), 0, 100000000000000000000);
+        }
+      } catch (e) {
+        console.log(
+          `Base percent: ${toBN(annualPercent)
+            .minus(1)
+            .multipliedBy(100)
+            .toString()}%. Max rate will be reached at ${y} month.`
+        );
+        return;
+      }
+    }
+
+    console.log(`Base percent: ${toBN(annualPercent).minus(1).multipliedBy(100).toString()}%. Done.`);
```

Comment:
console.log in prod code?

Reply:
Yep, it is very helpful to see when max compound rate willl be reached. For example: `Base percent: 10%. Max rate will be reached at 3345 month.`

Reply:
Yep, it is very helpful to see when max compound rate willl be reached. For example: `Base percent: 10%. Max rate will be reached at 3345 month.`

Reply:
Let's not include `console.log` in prod code. The successive users could include these in their own tests

Reply:
I don't see any problem, these are just tests. And the logs do not interfere with anything, but are part of the test. Namely, they indicate the duration of the contract work.

Reply:
Why do you need this in prod? Production has to clean AF and no console.logs should be there